File: code/platform/src/main/java/org/fbme/ide/platform/editor/HeaderedNodeEditor.java
package org.fbme.ide.platform.editor;

import com.intellij.openapi.actionSystem.*;
import com.intellij.openapi.editor.Document;
import jetbrains.mps.ide.editor.BaseNodeEditor;
import jetbrains.mps.ide.undo.MPSUndoUtil;
import jetbrains.mps.nodeEditor.NodeEditorComponent;
import jetbrains.mps.nodefs.MPSNodeVirtualFile;
import jetbrains.mps.nodefs.NodeVirtualFileSystem;
import jetbrains.mps.openapi.editor.EditorState;
import jetbrains.mps.openapi.editor.selection.SelectionManager;
import jetbrains.mps.project.Project;
import org.fbme.ide.iec61499.editor.ProjectEditorSpecs;
import org.jdom.Element;
import org.jdom.JDOMException;
import org.jdom.input.SAXBuilder;
import org.jetbrains.annotations.NonNls;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.jetbrains.mps.openapi.model.SNode;

import javax.swing.*;
import java.awt.*;
import java.io.IOException;
import java.io.StringReader;
import java.util.Collections;
import java.util.List;
import java.util.Objects;
import java.util.stream.Collectors;

public class HeaderedNodeEditor extends BaseNodeEditor {

    private final MPSNodeVirtualFile myVirtualFile;
    private final SNode myBaseNode;
    private final JPanel myHeaderPanel;

    private EditorProjection myCurrentProjection;
    private JComponent myProjectionComponent;

    private final ActionGroup myToolbarActionGroup;

    {
        final ProjectionChooserComboBoxAction action = new ProjectionChooserComboBoxAction(this);
        myToolbarActionGroup = new ActionGroup() {
            @NotNull
            @Override
            public AnAction[] getChildren(@Nullable AnActionEvent event) {
                return new AnAction[]{action};
            }
        };
    }

    private final ActionToolbar myChooserToolbar = ActionManager.getInstance().createActionToolbar(ActionPlaces.TOOLBAR, myToolbarActionGroup, true);

    public HeaderedNodeEditor(SNode baseNode, @NotNull Project mpsProject) {
        super(mpsProject);
        myBaseNode = baseNode;
        myVirtualFile = NodeVirtualFileSystem.getInstance().getFileFor(mpsProject.getRepository(), myBaseNode);

        myHeaderPanel = new JPanel(new BorderLayout());

        JComponent component = myChooserToolbar.getComponent();
        myHeaderPanel.add(component, BorderLayout.EAST);
        getComponent().add(myHeaderPanel, BorderLayout.NORTH);
        editNode(myBaseNode.getReference(), null);

        initializeFirstAvailableProjection();
    }

    public JComponent getProjectionComponent() {
        return myProjectionComponent;
    }

    @Override
    public boolean isTabbed() {
        return true;
    }

    @Override
    public void dispose() {
        super.dispose();
    }

    @Override
    public void showNode(SNode node, boolean select) {
        NodeEditorComponent ec = getCurrentEditorComponent();
        ProjectEditorSpecs editorSpecs = ProjectEditorSpecs.getInstance(myProject);
        String projectState = editorSpecs.getSpecProjectionState(node);
        node = editorSpecs.getSpecTarget(node);
        if (projectState != null) {
            SAXBuilder builder = new SAXBuilder();
            try {
                Element state = builder.build(new StringReader(projectState)).getRootElement();
                if (state != null) {
                    restoreState(state);
                }
            } catch (JDOMException | IOException e) {
                throw new RuntimeException(e);
            }
        }

        if (select) {
            ec.selectNode(node);
        } else {
            ec.getSelectionManager().setSelection(node, SelectionManager.FOCUS_POLICY_CELL, 0);
        }
    }

    public void chooseProjection(@NotNull EditorProjection projection) {
        myCurrentProjection = projection;
        updateProjectionState();
    }

    @NotNull
    public EditorProjection getCurrentProjection() {
        return myCurrentProjection;
    }

    public void updateProjectionState() {
        final NodeEditorComponent component = getCurrentEditorComponent();
        component.getUpdater().setInitialEditorHints(myCurrentProjection.getInitialEditorHints());
        myProject.getModelAccess().runReadAction(component::rebuildEditorContent);
        if (myProjectionComponent != null) {
            myHeaderPanel.remove(myProjectionComponent);
        }
        myProjectionComponent = myCurrentProjection.createHeaderComponent();
        if (myProjectionComponent != null) {
            myHeaderPanel.add(myProjectionComponent, BorderLayout.WEST);
        }
        myChooserToolbar.updateActionsImmediately();
    }

    @Override
    public List<Document> getAllEditedDocuments() {
        final MPSNodeVirtualFile virtualFile = getCurrentEditorComponent().getVirtualFile();
        return virtualFile != null ? Collections.singletonList(MPSUndoUtil.getDoc(virtualFile)) : Collections.emptyList();
    }

    @Override
    public Object getData(@NonNls String dataId) {
        if (dataId.equals(LangDataKeys.VIRTUAL_FILE.getName())) {
            return myVirtualFile;
        }
        if (dataId.equals(HeaderedEditorDataKeys.EDITOR.getName())) {
            return this;
        }
        if (myCurrentProjection != null) {
            return myCurrentProjection.getData(dataId);
        }
        return null;
    }

    @Override
    public EditorState saveState() {
        HeaderedEditorState state = new HeaderedEditorState();
        saveState(state);
        return state;
    }

    @Override
    protected void saveState(BaseNodeEditor.BaseEditorState state) {
        super.saveState(state);
        if (state instanceof HeaderedEditorState && myCurrentProjection != null) {
            Element projectionState = createDefaultState(myCurrentProjection);
            myCurrentProjection.saveState(projectionState);
            ((HeaderedEditorState) state).myProjectionState = projectionState;
        }
    }

    @Override
    public void loadState(@NotNull final EditorState state) {
        if (state instanceof BaseNodeEditor.BaseEditorState) {
            final BaseNodeEditor.BaseEditorState realState = (BaseEditorState) state;
            myProject.getModelAccess().runReadAction(() -> {
                if (realState instanceof HeaderedEditorState) {
                    Element projectionState = ((HeaderedEditorState) realState).myProjectionState;
                    restoreState(projectionState);
                } else {
                    // regular editor was shown for that node last time
                    initializeFirstAvailableProjection();
                }
                Element componentState = new Element("container");
                realState.save(componentState);
                Element memento = componentState.getChild("memento");
                if (memento != null) {
                    memento.removeChildren("enabledHints");
                    Element hintsElement = new Element("enabledHints");
                    for (String hint : myCurrentProjection.getInitialEditorHints()) {
                        Element hintElement = new Element("enabledHintsElement");
                        hintElement.setAttribute("enabledHintsAttribute", hint);
                        hintsElement.addContent(hintElement);
                    }
                    memento.addContent(hintsElement);
                }
                realState.load(componentState);
            });
            super.loadState(realState);
        }
    }

    private void restoreState(Element projectionState) {
        final String controllerId = projectionState.getAttributeValue(CONTROLLER_ID_KEY);
        String projectionName = projectionState.getAttributeValue(PROJECTION_NAME_KEY);
        if (controllerId != null && projectionName != null) {
            EditorProjectionController controller = EditorProjectionControllerRegistry.getInstance()
                    .getFactories().stream()
                    .filter(it -> Objects.equals(it.getId(), controllerId))
                    .findFirst().orElseThrow()
                    .create(myBaseNode, myProject);
            chooseProjection(controller.restoreProjection(projectionName, projectionState));
        }
    }

    private void initializeFirstAvailableProjection() {
        List<EditorProjectionController.Factory> factories = EditorProjectionControllerRegistry.getInstance()
                .getFactories().stream()
                .filter(it -> it.isApplicable(myBaseNode))
                .collect(Collectors.toList());

        for (EditorProjectionController.Factory factory : factories) {
            EditorProjectionController controller = factory.create(myBaseNode, myProject);
            EditorProjection defaultProjection = controller.createDefaultProjection();
            if (defaultProjection != null) {
                chooseProjection(defaultProjection);
                return;
            }
        }
    }

    public static final String CONTROLLER_ID_KEY = "id";
    public static final String PROJECTION_NAME_KEY = "name";
    public static final String PROJECTION_DATA_KEY = "projection_data";

    private static class HeaderedEditorState extends BaseNodeEditor.BaseEditorState {

        private Element myProjectionState;

        @Override
        public void save(Element e) {
            e.addContent(myProjectionState);
        }

        @Override
        public void load(Element e) {
            myProjectionState = e.getChild(PROJECTION_DATA_KEY).clone();
        }
    }

    public static Element createDefaultState(@NotNull EditorProjection projection) {
        String controllerId = projection.getController().getId();
        String projectionName = projection.getName();

        Element element = new Element(PROJECTION_DATA_KEY);
        element.setAttribute(CONTROLLER_ID_KEY, controllerId);
        element.setAttribute(PROJECTION_NAME_KEY, projectionName);
        return element;
    }
}

File: code/platform/src/main/kotlin/org/fbme/ide/platform/editor/HeaderedNodeEditor.kt
package org.fbme.ide.platform.editor

import com.intellij.openapi.actionSystem.*
import com.intellij.openapi.editor.Document
import jetbrains.mps.ide.editor.BaseNodeEditor
import jetbrains.mps.ide.undo.MPSUndoUtil
import jetbrains.mps.nodefs.MPSNodeVirtualFile
import jetbrains.mps.nodefs.NodeVirtualFileSystem
import jetbrains.mps.openapi.editor.EditorState
import jetbrains.mps.openapi.editor.selection.SelectionManager
import jetbrains.mps.project.Project
import org.fbme.ide.iec61499.editor.ProjectEditorSpecs
import org.jdom.Element
import org.jdom.JDOMException
import org.jdom.input.SAXBuilder
import org.jetbrains.annotations.NonNls
import org.jetbrains.mps.openapi.model.SNode
import java.awt.BorderLayout
import java.io.IOException
import java.io.StringReader
import javax.swing.JComponent
import javax.swing.JPanel

class HeaderedNodeEditor(val baseNode: SNode, mpsProject: Project) : BaseNodeEditor(mpsProject) {
    private val myVirtualFile: MPSNodeVirtualFile
    private val headerPanel: JPanel
    private val headerInfoPanel = ProjectionHeaderInfoPanel(this, mpsProject)
    var currentProjection: EditorProjection? = null
        private set
    var projectionComponent: JComponent? = null
        private set

    init {
        myVirtualFile = NodeVirtualFileSystem.getInstance().getFileFor(mpsProject.repository, baseNode)
        headerPanel = JPanel(BorderLayout())
        headerPanel.add(headerInfoPanel, BorderLayout.WEST)
        headerPanel.add(ProjectionChooserPanel(baseNode, mpsProject), BorderLayout.EAST)
        getComponent().add(headerPanel, BorderLayout.NORTH)
        editNode(baseNode.reference, null)
        initializeFirstAvailableProjection()
    }

    override fun isTabbed(): Boolean {
        return true
    }

    override fun showNode(node: SNode, select: Boolean) {
        val ec = currentEditorComponent
        val editorSpecs = ProjectEditorSpecs.getInstance(myProject)
        val projectState = editorSpecs.getSpecProjectionState(node)
        val target = editorSpecs.getSpecTarget(node)
        if (projectState != null) {
            val builder = SAXBuilder()
            try {
                val state = builder.build(StringReader(projectState)).rootElement
                state?.let { restoreState(it) }
            } catch (e: JDOMException) {
                throw RuntimeException(e)
            } catch (e: IOException) {
                throw RuntimeException(e)
            }
        }
        if (select) {
            ec.selectNode(target)
        } else {
            ec.selectionManager.setSelection(target, SelectionManager.FOCUS_POLICY_CELL, 0)
        }
    }

    fun chooseProjection(projection: EditorProjection) {
        currentProjection = projection

        val component = currentEditorComponent
        component.updater.initialEditorHints = projection.initialEditorHints
        myProject.modelAccess.runReadAction { component.rebuildEditorContent() }

        headerInfoPanel.update()
    }

    override fun getAllEditedDocuments(): List<Document> {
        val virtualFile = currentEditorComponent.virtualFile
        return if (virtualFile != null) listOf(MPSUndoUtil.getDoc(virtualFile)) else emptyList()
    }

    public override fun getData(@NonNls dataId: String): Any? {
        if (dataId == LangDataKeys.VIRTUAL_FILE.name) {
            return myVirtualFile
        }
        if (dataId == HeaderedEditorDataKeys.EDITOR.name) {
            return this
        }
        return currentProjection?.getData(dataId)
    }

    override fun saveState(): EditorState {
        val state = HeaderedEditorState()
        saveState(state)
        return state
    }

    override fun saveState(state: BaseEditorState) {
        super.saveState(state)
        if (state is HeaderedEditorState && currentProjection != null) {
            val projectionState = createDefaultState(
                    currentProjection!!
            )
            currentProjection!!.saveState(projectionState)
            state.myProjectionState = projectionState
        }
    }

    override fun loadState(state: EditorState) {
        if (state is BaseEditorState) {
            myProject.modelAccess.runReadAction {
                if (state is HeaderedEditorState) {
                    val projectionState = state.myProjectionState
                    restoreState(projectionState)
                } else {
                    // regular editor was shown for that node last time
                    initializeFirstAvailableProjection()
                }
                val componentState = Element("container")
                state.save(componentState)
                val memento = componentState.getChild("memento")
                if (memento != null) {
                    memento.removeChildren("enabledHints")
                    val hintsElement = Element("enabledHints")
                    for (hint in currentProjection!!.initialEditorHints) {
                        val hintElement = Element("enabledHintsElement")
                        hintElement.setAttribute("enabledHintsAttribute", hint)
                        hintsElement.addContent(hintElement)
                    }
                    memento.addContent(hintsElement)
                }
                state.load(componentState)
            }
            super.loadState(state)
        }
    }

    private fun restoreState(projectionState: Element?) {
        val controllerId = projectionState!!.getAttributeValue(CONTROLLER_ID_KEY)
        val projectionName = projectionState.getAttributeValue(PROJECTION_NAME_KEY)
        if (controllerId != null && projectionName != null) {
            val controller: EditorProjectionController = EditorProjectionControllerRegistry.instance
                    .factories.stream()
                    .filter { it.id == controllerId }
                    .findFirst().orElseThrow()
                    .create(baseNode, myProject)
            chooseProjection(controller.restoreProjection(projectionName, projectionState)!!)
        }
    }

    private fun initializeFirstAvailableProjection() {
        val factories: List<EditorProjectionController.Factory> =
                EditorProjectionControllerRegistry.instance.factories.filter { it.isApplicable(baseNode) }
        for (factory in factories) {
            val controller = factory.create(baseNode, myProject)
            val defaultProjection = controller.createDefaultProjection()
            if (defaultProjection != null) {
                chooseProjection(defaultProjection)
                return
            }
        }
    }

    private class HeaderedEditorState : BaseEditorState() {
        var myProjectionState: Element? = null
        override fun save(e: Element) {
            e.addContent(myProjectionState!!.clone())
        }

        override fun load(e: Element) {
            myProjectionState = e.getChild(PROJECTION_DATA_KEY).clone()
        }
    }

    companion object {
        const val CONTROLLER_ID_KEY = "id"
        const val PROJECTION_NAME_KEY = "name"
        const val PROJECTION_DATA_KEY = "projection_data"
        fun createDefaultState(projection: EditorProjection) =
                createState(projection.controller.id, projection.name)

        fun createState(controllerId: String, projectionName: String): Element {
            val element = Element(PROJECTION_DATA_KEY)
            element.setAttribute(CONTROLLER_ID_KEY, controllerId)
            element.setAttribute(PROJECTION_NAME_KEY, projectionName)
            return element
        }
    }
}
